pkgs/by-name/ri/ringboard/package.nix

```diff
@@ -0,0 +1,96 @@
+{
+  stdenv,
+  rustPlatform,
+  fetchFromGitHub,
+  lib,
+  libxkbcommon,
+  libGL,
+  wayland,
+  xorg,
+  makeWrapper,
+}:
+
+rustPlatform.buildRustPackage (finalAttrs: {
+  pname = "ringboard";
+
+  # release version needs nightly, so we use a custom tree, see:
+  # https://github.com/SUPERCILEX/clipboard-history/issues/22#issuecomment-3322075172
+  version = "0.12.2-unstable-2025-09-23";
+
+  src = fetchFromGitHub {
+    owner = "SUPERCILEX";
+    repo = "clipboard-history";
+    rev = "228a39dd8a9aece0bb06f68ad44906b297270628";
+    hash = "sha256-qA7wwvWnnZHN9edkmubEo37F+peU0LQGo/Zl8FpywuE=";
+  };
+
+  cargoHash = "sha256-MFfuUu/hpb6Uaqe21bvXNKRyJazAL5m+Vw/vAeeDYEk=";
+
+  nativeBuildInputs = [
+    makeWrapper
+  ];
+
+  buildInputs = [
+    libxkbcommon
+    libGL
+
+    # WINIT_UNIX_BACKEND=wayland
+    wayland
+
+    # WINIT_UNIX_BACKEND=x11
+    xorg.libXcursor
+    xorg.libXrandr
+    xorg.libXi
+    xorg.libX11
+  ];
+
+  buildPhase = ''
+    runHook preBuild
+
+    local flagsArray=("-j $NIX_BUILD_CORES --target ${stdenv.hostPlatform.rust.rustcTarget} --offline --release");
+    concatTo flagsArray cargoBuildFlags;
+
+    echo "Building package: clipboard-history-server"
+    cargo build $flagsArray --package clipboard-history-server --no-default-features --features systemd
+    echo "Building package: clipboard-history-x11"
+    cargo build $flagsArray --package clipboard-history-x11 --no-default-features
+    echo "Building package: clipboard-history-wayland"
+    cargo build $flagsArray --package clipboard-history-wayland --no-default-features
+    echo "Building package: clipboard-history"
+    cargo build $flagsArray --package clipboard-history
+    echo "Building package: clipboard-history-tui"
+    cargo build $flagsArray --package clipboard-history-tui
+    echo "Building package: clipboard-history-egui"
+    cargo build $flagsArray --package clipboard-history-egui
```

Comment:
This should have ` --no-default-features --features $XDG_SESSION_TYPE`. But maybe you can't do that in nix? The persistent background window only works with `--no-default-features --features x11`, so you need to pass in x11 or wayland depending on what you're running on.

Also it's a little sad that everybody needs to install both wayland and x11 deps. So how hard would it be to split this up into two packages? One for x11 people and one for wayland people.

Reply:
This is doable. We can add an argument to the package args up top called `displayServer` and conditionally include flags and deps, and then register both variants in `all-packages.nix`.

Reply:
Yes. I don't have a computer atm, but see for example https://github.com/NixOS/nixpkgs/blob/e57b3b16ad8758fd681511a078f35c416a8cc939/pkgs/tools/security/pass/rofi-pass.nix

Here they have an argument named `backend` which should be either `"x11"` or `"wayland"`, and use it further down in the package to conditionally include `buildInputs` or run stuff in the phases.

After adding this argument, you can declare named variants of the package in `pkgs/top-level/all-packages.nix`, something like

```nix
    ringboard-wayland = callPackage ../by-name/ri/ringboard.nix { backend = "wayland"; };
ringboard-x11 = callPackage ../by-name/ri/ringboard.nix { backend = "x11"; };
```

Not sure if `backend` is a good name for the argument, but hopefully you get the idea.

Reply:
Maybe call it `displayServer`? But `backend` seems reasonable too.

Reply:
@SUPERCILEX I'll try and remove the script from the service now.
@h7x4 Got any tips on a clean architecture for that?

Reply:
Isn't it hallucinating `config.programs.wayland.enable`? Also ideally you wouldn't include the other service at all rather than disabling it.

Reply:
Sorry, can't you play the same trick as your other commit and pass in `${displayServer}` to the relevant places so a single service with the right names gets generated?

Reply:
Oh, duh, I should have checked that, sorry!
It's bed-time for me, hopefully @h7x4 will have outlined a better strategy when I wake up.

Reply:
Even though the packages now only contains either wayland or x11 binaries, the module wouldn't know which one to use.

Maybe it's good enough to add an `mkPackageOption` so the user can change which package they want to use themselves? Alternatively, we could make the default of the package option check if wayland is enabled in `config`, and choose a package based on that.

I don't think we need a separate service for each variant.

I'm on mobile atm, will write a diff later. Ping me if I've been silent for a day, then I've likely just forgot :)

Reply:
@h7x4 
> Ping me if I've been silent for a day, then I've likely just forgot :)

8 hours is a workday, right?  (if that's not what you meant: sorry for being impatient!)